### PR TITLE
Add DESCRIBE OUTPUT support

### PR DESCRIPTION
**Pull request by rschlussel**
We have docs for this too.  Should I add it to #6291 or wait for that to get merged and add the docs for this later?

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+            Row[] rows = analysis.getRootScope().getRelationType().getVisibleFields().stream().map(field -> createDescribeOutputRow(field, analysis)).toArray(Row[]::new);
+            Query query = simpleQuery(
+                    selectList(nameReference("Column Name"),
+                            nameReference("Table"),
+                            nameReference("Schema"),
+                            nameReference("Connector"),
+                            nameReference("Type"),
+                            nameReference("Type Size"),
+                            nameReference("Aliased"),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: Does this come from the SQL spec? What does it indicate?

**Comment by martint** [comment]
I was looking at the spec and realized that `DESCRIBE OUTPUT` is supposed to be the same as `DESCRIBE`. Is the `DESCRIBE` in Presto even close to matching what the spec dictates (or at least, not preclude it?):

```
<describe output statement> ::=
  DESCRIBE [ OUTPUT ] <described object> <using descriptor> [ <nesting option> ]
```

Have you looked into what `using descriptor` and SQL descriptor areas are all about?

**Comment by rschlussel** [comment]
The `using descriptor` relates specifically to embedded sql, so it's not relevant to us (they're described in section 20.1 of the spec though).  All of the syntax for prepared statements in the sql spec is for use in embedded sql.  We just borrowed it.

Presto's use of `DESCRIBE` as `DESCRIBE <table>` is not a part of the sql spec.  Also, we require the word `OUTPUT` if you're trying to describe a prepared statement rather than a table (since `DESCRIBE` without a qualifier is already taken for describing a table).   But the `DESCRIBE` statement in the spec is for embedded sql, so both features are really an extension of the spec.

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java (hunk 1 above)
Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: Yes this was a requirement from Simba. They want to know if a column is aliased and if that column can be traced to a base table, what table it originates from.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: I'm not clear about the semantics of this. What's the expected output if the query has a "join ... using", union/intersect/except, etc., and a column is derived from multiple columns across different tables?

Comment (petroav) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: In the case of this join `SELECT * FROM nation AS n JOIN region AS r ON n.regionkey = r.regionkey` we're able to track all the columns so we display all the information:

```
 Column Name | Table  | Schema | Connector |     Type     | Type Size | Aliased | Row Count Query 
-------------+--------+--------+-----------+--------------+-----------+---------+-----------------
 nationkey   | nation | sf1    | tpch      | bigint       |         8 | false   | false           
 name        | nation | sf1    | tpch      | varchar(25)  |         0 | false   | false           
 regionkey   | nation | sf1    | tpch      | bigint       |         8 | false   | false           
 comment     | nation | sf1    | tpch      | varchar(152) |         0 | false   | false           
 regionkey   | region | sf1    | tpch      | bigint       |         8 | false   | false           
 name        | region | sf1    | tpch      | varchar(25)  |         0 | false   | false           
 comment     | region | sf1    | tpch      | varchar(152) |         0 | false   | false
```

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: Why treat "row count queries" differently? Why show a single column of type, say, bigint? This is both more general and future-proof, e.g., if we decide to add other columns to the output of certain commands.

Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: Simba had requested this for the driver so they would know whether they would need to retrieve metadata. But I spoke to them and they said that returning no rows when there is no data output would work for them as well, so I'll make that change.

**Comment by rschlussel** [comment]
Addressed comments

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+            String columnName;
+            if (field.getName().isPresent()) {
+                columnName = field.getName().get();
+            }
+            else {
+                int columnIndex = ImmutableList.copyOf(analysis.getOutputDescriptor().getVisibleFields()).indexOf(field);
+                columnName = "_col" + columnIndex;
+            }
+
+            Optional<QualifiedObjectName> qualifiedOriginTable = field.getOriginTable();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: Rename this variable to `originTable`

File: presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+            Optional<String> limit = Optional.empty();
+            Row[] rows = analysis.getRootScope().getRelationType().getVisibleFields().stream().map(field -> createDescribeOutputRow(field, analysis)).toArray(Row[]::new);
+            if (rows.length == 0) {
+                NullLiteral nullLiteral = new NullLiteral();
+                rows = new Row[] {row(nullLiteral, nullLiteral, nullLiteral, nullLiteral, nullLiteral, nullLiteral, nullLiteral)};
+                limit = Optional.of("0");
+            }
+            return simpleQuery(
+                    selectList(nameReference("Column Name"),
+                            nameReference("Table"),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: Would it make more sense to order the fields as catalog, schema, table?

Comment (rschlussel) [inline presto-main/src/main/java/com/facebook/presto/sql/rewrite/DescribeOutputRewrite.java]: Sure, that makes sense

**Comment by martint** [comment]
Merged, thanks!

**Pull request closed without merging**